be/src/block_cache/disk_space_monitor.cpp

```diff
@@ -0,0 +1,259 @@
+// Copyright 2021-present StarRocks, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "block_cache/disk_space_monitor.h"
+
+#include "block_cache/block_cache.h"
+#include "common/config.h"
+#include "util/thread.h"
+
+namespace starrocks {
+
+#ifndef BE_TEST
+const size_t DiskSpaceMonitor::QUOTA_ALIGN_UNIT = 10uL * 1024 * 1024 * 1024;
+#else
+const size_t DiskSpaceMonitor::QUOTA_ALIGN_UNIT = 10uL * 1024 * 1024;
+#endif
+
+const int64_t DiskSpaceMonitor::AUTO_INCREASE_THRESHOLD = 90;
+
+StatusOr<size_t> DiskSpaceMonitor::FileSystemWrapper::directory_capacity(const std::string& dir) {
+    size_t capacity = 0;
+    auto st = FileSystem::Default()->iterate_dir2(dir, [&](DirEntry entry) {
+        capacity += entry.size.value();
+        return true;
+    });
+    RETURN_IF_ERROR(st);
+    return capacity;
+}
+
+DiskSpaceMonitor::DiskSpaceMonitor(BlockCache* cache) : _fs(std::make_unique<FileSystemWrapper>()), _cache(cache) {}
+
+DiskSpaceMonitor::~DiskSpaceMonitor() {
+    stop();
+}
+
+void DiskSpaceMonitor::start() {
+    std::unique_lock<std::mutex> lck(_mutex);
+    if (!_stopped.load(std::memory_order_acquire)) {
+        return;
+    }
+    _stopped.store(false, std::memory_order_release);
+    _adjust_datacache_thread = std::thread([this] { _adjust_datacache_callback(); });
+    Thread::set_thread_name(_adjust_datacache_thread, "adjust_datacache");
+}
+
+void DiskSpaceMonitor::stop() {
+    if (_stopped.load(std::memory_order_acquire)) {
+        return;
+    }
+    _stopped.store(true, std::memory_order_release);
+    _cv.notify_one();
+    if (_adjust_datacache_thread.joinable()) {
+        _adjust_datacache_thread.join();
+    }
+}
+
+void DiskSpaceMonitor::_adjust_datacache_callback() {
+    while (!_stopped.load(std::memory_order_acquire)) {
+        std::unique_lock<std::mutex> lck(_mutex);
+        if (config::datacache_enable && config::datacache_auto_adjust_enable &&
+            !_adjusting.load(std::memory_order_acquire)) {
+            _update_disk_stats();
+            _update_cache_stats();
+            if (_adjust_spaces_by_disk_usage(false)) {
+                Status st = adjust_cache_quota(_dir_spaces);
+                LOG_IF(WARNING, !st.ok()) << "fail to adjust datacache disk quota, reason: " << st.message();
+            }
+        }
+        _cv.wait_for(lck, std::chrono::seconds(config::datacache_disk_adjust_interval_seconds));
```

Comment:
this wait is too long for a quick BE/CN exit. (default to 10s)

Reply:
stop may hang forever.

Consider following execution order
```
1. [callback thread] line 69, check false, execute to line 70
2. [main thread] execute line 61, 62, 63, wait for callback thread join
3. [callback thread] execute to line 80, wait for cv notify.
```

Reply:
> stop may hang forever.
> 
> Consider following execution order
> 
> ```
> 1. [callback thread] line 69, check false, execute to line 70
> 2. [main thread] execute line 61, 62, 63, wait for callback thread join
> 3. [callback thread] execute to line 80, wait for cv notify.
> ```

Thanks a lot! Changed it.